src/mds/MDSAuthCaps.h

```diff
@@ -29,53 +29,67 @@ struct MDSCapSpec {
   MDSCapSpec() : read(false), write(false), any(false) {}
   MDSCapSpec(bool r_, bool w_, bool a_) : read(r_), write(w_), any(a_) {}
 
-  bool allow_all() const {return any;}
+  bool allow_all() const {
+    return any;
+  }
+  bool allows(bool r, bool w) const {
+    if (any)
+      return true;
+    if (r && !read)
+      return false;
+    if (w && !write)
+      return false;
+    return true;
+  }
 };
 
+// conditions before we are allowed to do it
 struct MDSCapMatch {
   static const int MDS_AUTH_UID_ANY = -1;
   static const std::string MDS_AUTH_PATH_ROOT;
 
-  int uid;  // Require UID to be equal to this, if !=MDS_AUTH_UID_ANY
+  int uid;           // Require UID to be equal to this, if !=MDS_AUTH_UID_ANY
+  std::vector<int> gids;  // Use these GIDs
   std::string path;  // Require path to be child of this (may be "/" for any)
 
   MDSCapMatch() : uid(MDS_AUTH_UID_ANY), path(MDS_AUTH_PATH_ROOT) {}
-  MDSCapMatch(int uid_) : uid(uid_), path(MDS_AUTH_PATH_ROOT) {}
+  MDSCapMatch(int uid_, std::vector<int>& gids_)
+    : uid(uid_), gids(gids_), path(MDS_AUTH_PATH_ROOT) {}
   MDSCapMatch(std::string path_) : uid(MDS_AUTH_UID_ANY), path(path_) {}
-  MDSCapMatch(std::string path_, int uid_) : uid(uid_), path(path_) {}
+  MDSCapMatch(std::string path_, int uid_, std::vector<int>& gids_)
+    : uid(uid_), gids(gids_), path(path_) {}
   
   bool is_match_all() const
   {
     return uid == MDS_AUTH_UID_ANY && path == "/";
   }
-
-  bool match(const std::string &target_path, const int target_uid) const {
-    return (target_path.find(path) == 0 && (target_uid == uid || uid == MDS_AUTH_UID_ANY)); 
-  }
+  bool match(const std::string &target_path, const int target_uid) const;
 };
 
 struct MDSCapGrant {
   MDSCapSpec spec;
   MDSCapMatch match;
 
-  MDSCapGrant(const MDSCapSpec &spec_, const MDSCapMatch &match_) : spec(spec_), match(match_) {}
+  MDSCapGrant(const MDSCapSpec &spec_, const MDSCapMatch &match_)
+    : spec(spec_), match(match_) {}
   MDSCapGrant() {}
 };
 
-class MDSAuthCaps
+struct MDSAuthCaps
```

Comment:
It's so that Server.cc can iterate over the list of grants.  I'm happy 
exposing the idea that a cap is a list of grants...